### PR TITLE
Added Core Peer Forwarder decorator and interface

### LINKED ISSUE #1589: Add PeerForwardingProcessingDecorator skeleton
Add PeerForwardingProcessingDecorator to Data Prepper core which implements processor.

### LINKED ISSUE #1590: Add RequiresPeerForwarding interface
Add RequiresPeerForwarding interface which is implemented by stateful processors. Data Prepper will detect this plugins which implement it and configure peer forwarder for these processors.

### PR DESCRIPTION
**Pull request by asifsmohammed**
Signed-off-by: Asif Sohail Mohammed <nsifmoh@amazon.com>

### Description
Adds core peer forwarder decorator skeleton and interface
 
### Issues Resolved
Resolves #1589 
Resolves #1590 
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwardingProcessingDecorator.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import com.amazon.dataprepper.model.event.Event;
+import com.amazon.dataprepper.model.processor.Processor;
+import com.amazon.dataprepper.model.record.Record;
+import org.apache.commons.lang3.NotImplementedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+
+public class PeerForwardingProcessingDecorator implements Processor<Record<Event>, Record<Event>> {
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwardingProcessingDecorator.java:17]: Discussed offline: we can push back this until adding metrics

**Review by chenqi0805 — DISMISSED** [review]

File: data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwardingProcessingDecorator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (asifsmohammed) [reply data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwardingProcessingDecorator.java:17]: We need to think about how we can pass `PluginMetrics` to constructor of `PeerForwardingProcessorDecorator`

File: data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwardingProcessorDecorator.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Collection;
+
+public class PeerForwardingProcessorDecorator implements Processor<Record<Event>, Record<Event>> {
+    private static final Logger LOG = LoggerFactory.getLogger(PeerForwardingProcessorDecorator.class);
+
+
+    @Override
+    public Collection<Record<Event>> execute(final Collection<Record<Event>> collection) {
+        LOG.info("Peer Forwarder not implemented yet, skipping events for now.");
+        throw new NotImplementedException();
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwardingProcessorDecorator.java:24]: Java has a built-in exception you can use for now: `UnsupportedOperationException`. You can remove the Apache dependency by making this change.

File: settings.gradle [hunk 3]
```diff
@@ -59,3 +59,4 @@ include 'data-prepper-plugins:date-processor'
 include 'data-prepper-expression'
 include 'data-prepper-plugins:mutate-string-processors'
 include 'data-prepper-plugins:s3-source'
+include 'core-peer-forwarder'
```
Comment (dlvenable) [inline settings.gradle:62]: You added a new project here. But, I don't see any project files under `core-peer-forwarder`. We can just remove this for now.

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by chenqi0805 — DISMISSED** [review]

**Pull request merged**